browser_patches/webkit/patches/bootstrap.diff

```diff
@@ -12380,7 +13461,53 @@ index 0ed7fc66150a924d0caa4b0d98d8e6cc6208ca38..fba5cd0a8e91557dbc8d551cbfe980fe
      if (flagsToUpdate & ActivityState::IsFocused && pageClient().isViewFocused())
          m_activityState.add(ActivityState::IsFocused);
      if (flagsToUpdate & ActivityState::WindowIsActive && pageClient().isViewWindowActive())
-@@ -2840,7 +2900,7 @@ static TrackingType mergeTrackingTypes(TrackingType a, TrackingType b)
+@@ -2434,11 +2494,17 @@ void WebPageProxy::didPerformDragControllerAction(Optional<WebCore::DragOperatio
+     pageClient().didPerformDragControllerAction();
+ }
+ 
+-#if PLATFORM(GTK)
++#if PLATFORM(GTK) || PLATFORM(WPE)
+ void WebPageProxy::startDrag(SelectionData&& selectionData, OptionSet<WebCore::DragOperation> dragOperationMask, const ShareableBitmap::Handle& dragImageHandle)
+ {
+-    RefPtr<ShareableBitmap> dragImage = !dragImageHandle.isNull() ? ShareableBitmap::create(dragImageHandle) : nullptr;
+-    pageClient().startDrag(WTFMove(selectionData), dragOperationMask, WTFMove(dragImage));
++    if (m_interceptDrags) {
++        m_dragSelectionData = WTFMove(selectionData);
++    } else {
++#if PLATFORM(GTK)
++        RefPtr<ShareableBitmap> dragImage = !dragImageHandle.isNull() ? ShareableBitmap::create(dragImageHandle) : nullptr;
++        pageClient().startDrag(WTFMove(selectionData), dragOperationMask, WTFMove(dragImage));
++#endif
++    }
+ 
+     didStartDrag();
+ }
+@@ -2570,8 +2636,22 @@ void WebPageProxy::processNextQueuedMouseEvent()
+         m_process->startResponsivenessTimer();
+     }
+ 
+-    LOG(MouseHandling, "UIProcess: sent mouse event %s (queue size %zu)", webMouseEventTypeString(eventType), m_mouseEventQueue.size());
+-    send(Messages::WebPage::MouseEvent(event));
++    if (!m_dragSelectionData) {
++        LOG(MouseHandling, "UIProcess: sent mouse event %s (queue size %zu)", webMouseEventTypeString(eventType), m_mouseEventQueue.size());
++        send(Messages::WebPage::MouseEvent(event));
++    } else {
++        DragData dragData(&*m_dragSelectionData, event.position(), event.globalPosition(), m_currentDragOperation);
```

Comment:
This logic seems out of place, why can't it be handled in the client similar to GTK?

Reply:
This patch has two parts. One part adds drag support to the web process for WPE. The second part is about intercepting drags in WebPageProxy before they hit the OS. In a follow up, I will move GTK to go down this code path as well. WebPageProxy essentially acts as its own drag client here, converting mouse events into drag events.